source/MaterialXRuntime/RtNode.cpp

```diff
@@ -16,100 +16,56 @@ namespace MaterialX
-        RtToken("type"),
-        RtToken("Value"),
-        RtToken("nodename"),
-        RtToken("nodegraph"),
-        RtToken("output"),
-        RtToken("member"),
-        RtToken("channels"),
-        RtToken("unit"),
-        RtToken("unittype")
-    };
-
-    static const RtTokenVec PUBLIC_INPUT_METADATA_NAMES
-    {
-        RtToken("name"),
-        RtToken("type"),
-        RtToken("Value"),
-        RtToken("nodename"),
-        RtToken("nodegraph"),
-        RtToken("output"),
-        RtToken("member"),
-        RtToken("channels")
-    };
-
-    static const RtTokenVec PUBLIC_OUTPUT_COLOR_METADATA_NAMES
-    {
-        RtToken("name"),
-        RtToken("type"),
-        RtToken("nodename"),
-        RtToken("output"),
-        RtToken("member"),
-        RtToken("colorspace"),
-        RtToken("width"),
-        RtToken("height"),
-        RtToken("bitdepth")
-    };
-    static const RtTokenVec PUBLIC_OUTPUT_METADATA_NAMES
+    // TODO: We should derive this from a data driven XML schema.
+    class NodePrimSpec : public PvtPrimSpec
     {
-        RtToken("name"),
-        RtToken("type"),
-        RtToken("nodename"),
-        RtToken("output"),
-        RtToken("member"),
-        RtToken("width"),
-        RtToken("height"),
-        RtToken("bitdepth")
-    };
-
-    static const RtTokenVec PUBLIC_METADATA_NAMES
-    {
-        RtToken("name"),
-        RtToken("type"),
-        RtToken("value"),
-        RtToken("doc"),
-        RtToken("xpos"),
-        RtToken("ypos"),
-        RtToken("width"),
-        RtToken("height"),
-        RtToken("uicolor"),
-        RtToken("uiname"),
-        RtToken("uivisible"),
-        RtToken("uiadvanced"),
-        RtToken("version"),
-        RtToken("cms"),
-        RtToken("cmsconfig"),
-        RtToken("colorspace"),
-        RtToken("namespace")
+    public:
+        NodePrimSpec()
+        {
+            addPrimAttribute(Tokens::DOC, RtType::STRING);
+            addPrimAttribute(Tokens::XPOS, RtType::FLOAT);
+            addPrimAttribute(Tokens::YPOS, RtType::FLOAT);
+            addPrimAttribute(Tokens::WIDTH, RtType::INTEGER);
+            addPrimAttribute(Tokens::HEIGHT, RtType::INTEGER);
+            addPrimAttribute(Tokens::UICOLOR, RtType::COLOR3);
+            addPrimAttribute(Tokens::UINAME, RtType::STRING);
+            addPrimAttribute(Tokens::VERSION, RtType::TOKEN);
+
+            addInputAttribute(Tokens::DOC, RtType::STRING);
+            addInputAttribute(Tokens::MEMBER, RtType::STRING);
+            addInputAttribute(Tokens::CHANNELS, RtType::STRING);
+            addInputAttribute(Tokens::UIADVANCED, RtType::BOOLEAN);
+            addInputAttribute(Tokens::UIVISIBLE, RtType::BOOLEAN);
+
+            addInputAttributeByType(RtType::COLOR3, Tokens::COLORSPACE, RtType::TOKEN);
+            addInputAttributeByType(RtType::COLOR4, Tokens::COLORSPACE, RtType::TOKEN);
+
+            addInputAttributeByType(RtType::FLOAT, Tokens::UNIT, RtType::TOKEN);
```

Comment:
Instead of having to repeat the same information with a different type, would it be better to take in a vector of types?

Reply:
This is temporary code since these constructors will read from XML schema files instead later, when we get to that. So I think we can leave it as is for now.